action/flow/instance/auditevents.go

```diff
@@ -0,0 +1,245 @@
+	return convertTaskStatus(te.taskInstance.status)
+}
+
+// Returns application name
+func (te *TaskEventContext) AppName() string {
+	return app.GetName()
+}
+
+// Returns application version
+func (te *TaskEventContext) AppVersion() string {
+	return app.GetVersion()
+}
+
+// Returns working data of current instance. e.g. key and value of current iteration for iterator task.
+func (te *TaskEventContext) GetWorkingData() map[string]interface{} {
+	attrs := make(map[string]interface{})
+	if te.taskInstance.HasWorkingData() {
+		for name, value := range te.taskInstance.workingData {
+			attrs[name] = value.Value()
+		}
+	}
+	return attrs
+}
+
+// Returns activity input data
+func (te *TaskEventContext) ActivityInput() map[string]interface{} {
+	attrs := make(map[string]interface{})
+	 if te.taskInstance.task.ActivityConfig().GetInputAttrs() != nil && te.taskInstance.inScope != nil {
+		 for name := range te.taskInstance.task.ActivityConfig().GetInputAttrs() {
+		 	inVal, _ := te.taskInstance.inScope.GetAttr(name)
+		 	attrs[name] = inVal
+		 }
+	 }
+	return attrs
+}
+
+// Returns output data for completed activity
+func (te *TaskEventContext) ActivityOutput() map[string]interface{} {
+	attrs := make(map[string]interface{})
+	if te.Status() == Completed && te.taskInstance.task.ActivityConfig().GetOutputAttrs() != nil && te.taskInstance.outScope != nil {
+		for name := range te.taskInstance.task.ActivityConfig().GetOutputAttrs() {
+			outVal, _ := te.taskInstance.outScope.GetAttr(name)
+			attrs[name] = outVal
+		}
+	}
+	return attrs
+}
+
+// Returns error for failed task
+func (te *TaskEventContext) Error() error {
+	return te.taskInstance.returnError
+}
+
+func convertFlowStatus(code model.FlowStatus) Status {
+	switch code {
+	case model.FlowStatusNotStarted:
+		return Created
+	case model.FlowStatusActive:
+		return Started
+	case model.FlowStatusCancelled:
+		return Cancelled
+	case model.FlowStatusCompleted:
+		return Completed
+	case model.FlowStatusFailed:
+		return Failed
+	}
+	return Unknown
+}
+
+func convertTaskStatus(code model.TaskStatus) Status {
+	switch code {
+	case model.TaskStatusNotStarted:
+		return Created
+	case model.TaskStatusEntered:
+		return Scheduled
+	case model.TaskStatusSkipped:
+		return Skipped
+	case model.TaskStatusReady:
+		return Started
+	case model.TaskStatusFailed:
+		return Failed
+	case model.TaskStatusDone:
+		return Completed
+	case model.TaskStatusWaiting:
+		return Waiting
+	}
+	return Unknown
+}
+
+// Registers listener for flow events
+func RegisterFlowEventListener(fel FlowEventListenerFunc) {
+	lock.Lock()
+	defer lock.Unlock()
+	flowEventListeners = append(flowEventListeners, fel)
+}
+
+// Registers listener for task events
+func RegisterTaskEventListener(tel TaskEventListenerFunc) {
+	lock.Lock()
+	defer lock.Unlock()
+	taskEventListeners = append(taskEventListeners, tel)
+}
+
+func publishFlowEvent(fe *FlowEventContext) {
+	for _, fel := range flowEventListeners {
+		go fel(fe)
```

Comment:
And at last, we should enable or disable with a parameter at registration-time whether the listener callback should be called in a goroutine or not. In fact our analysis concluded that the first event of a flow ("Created") is kind of special since it can be used to init objects that must be available to the children tasks events.

Reply:
So do you think its important to have the option on all event publishes or just on the flow "created"?

Reply:
@debovema : I am thinking on an alternate approach to ensure event ordering which I believe is basic principle of this mechanism. Lets discuss this again once I do that. :-)